src/password.migrate.js

```diff
@@ -0,0 +1,79 @@
+import { post } from "./api.js";
+import { setCookiesAndTokens } from "./cookies.js";
+import { store } from "./store.js";
+import { handleRedirect } from "./url.js";
+import { throwFormattedError } from "./utils.js";
+import { exchange } from "./refresh.js";
+import { getMfaHeaders, handleMfaRequired } from "./authentication.js";
+import { getPkceRequestQueryParams, redirectWithPkce } from "./pkce.js";
+
+/**
+ * Log a user in with email/username and password using the password/migrate endpoint.
+ * Redirect the browser after successful login based on the redirectTo value returned.
+ * @param {Object} params
+ * @param {string} params.email The user's email. One of email/username/emailOrUsername should be present.
+ * @param {string} params.username The user's username. One of email/username/emailOrUsername should be present.
+ * @param {string} params.emailOrUsername Either the user's email or username. One of email/username/emailOrUsername should be present.
+ * @param {string} params.password
+ * @param {string | boolean} params.redirect
+ *  URL to redirect to after login, or false to suppress redirect. Otherwise, redirects to the after-login path set on the server.
+ * @param {function} params.handleUpstreamResponse Function to run after receiving response, but before redirection
+ * @param {object} params.options
+ * @param {boolean} params.options.noResetEmail
+ *  By default, Userfront sends a password reset email if a user without a password tries to log in with a password.
+ *  Set options.noResetEmail = true to override this behavior and return an error instead.
+ *
+ */
+export async function loginWithPasswordMigrate({
+  email,
+  username,
+  emailOrUsername,
+  password,
+  redirect,
+  handleUpstreamResponse,
+  options,
+}) {
+  try {
+    const body = {
+      tenantId: store.tenantId,
+      emailOrUsername: email || username || emailOrUsername,
+      password,
+    };
+    if (options && options.noResetEmail) {
+      body.options = {
+        noResetEmail: true,
+      };
+    }
+
+    // Make the request to password/migrate
+    const { data } = await post(`/auth/password/migrate`, body, {
+      headers: getMfaHeaders(),
```

Comment:
**For present consideration:**

Note that `getMfaHeaders()` is always going to return an empty object in this case, since `password-migrate` is always a first factor. I support keeping it there as boilerplate to match all the other methods, just wanted to note it.

I noticed that there's no `mfaRequired` response handler below. How should we handle a situation where someone uses password migration with MFA Required set in their default auth flow? Is `password-migrate` even part of the auth flow system? (Could this be a situation where we might actually get use out of having multiple auth flows for a tenant, so they can migrate single-factor password users transparently but enforce MFA for new signups?)

I suspect we'll enable the MFA toggle & related features in the dashboard before revisiting migration. Will we want a tenant-specific flag for the dashboard to prevent tenants in active migration (i.e. the current customer specifically) from toggling on MFA Required, and messaging to explain the situation? Similarly, will some part of the system (API, I guess?) prevent a user with a `password-migrate` first factor being assigned a second factor on the user record?

(I think those are two separate use cases, and we could enable user specific opt-in MFA + add `mfaRequired` response handling here, without allowing MFA Required at the tenant level. So a user could migrate and then later enable a second factor, without being required to setup a second factor at the moment of migration.)

**For future thought:**

How does migration interact with existing MFA? Will it be possible to migrate supported MFA flows? What happens when a tenant that already has MFA Required in its default auth flow starts a password migration?

Reply:
Don't feel obligated to answer each question individually! Just want to make sure we know how migration + MFA is intended to work now and note any relevant work items that need to be completed before rolling out the MFA feature.

Reply:
These are all great questions, and I will just say for now that we don't have migration with MFA enabled fully figured out just yet! My first step is to get migration working without MFA, then with MFA enabled on the upstream service, then with MFA enabled on both.